### PR TITLE
show job logs (#643)

### PR DESCRIPTION
**Pull request by dgkanatsios**
Implementation for #643 
Specifically, added `streamLogs` boolean variable to Job object on brigade.js. This variable triggers the streaming/follow of the Job Pod(s) Logs, similar to `kubectl logs PODNAME -f` functionality.

### CODE REVIEW

**Comment by vdice** [comment]
@dgkanatsios code LGTM and tested with a local project.

Note: While logs indeed show up at the end of a job run, I'm not seeing the logs stream in realtime.  If streaming was one of the criteria/intended enhancements, perhaps some more work is needed -- otherwise, looks great!

**Comment by dgkanatsios** [comment]
thanks @vdice. As usual, you are correct, streaming (or in the K8s terminology, "following") the logs makes much more sense. Just pushed an implementation for that, let me know

**Comment by dgkanatsios** [comment]
Both done, thanks. The "done" output is a bit tricky to remove, since it's not being sent from the worker but from the [delegated runner](https://github.com/Azure/brigade/blob/master/pkg/script/delegated_runner.go#L217)

File: brigade-worker/src/job.ts [hunk 1]
```diff
@@ -245,6 +245,9 @@ export abstract class Job {
     return this._podName;
   }
 
+  /** displayLogs controls whether logs from the job Pod will be sent to output */
+  public displayLogs : boolean = false;
```
Comment (radu-matei) [inline brigade-worker/src/job.ts]: Is there a different variable name for following / tailing the logs?

I think one of those might be a little more intuitive if this actually follows logs (but this is just my opinion, though)

Comment (dgkanatsios) [reply brigade-worker/src/job.ts]: Hm, in the ‘brig run’ context, logs are streamed/followed/sent to stdout as ‘brig run’ is executed. There is no other behavior (mainly because ‘brig run’ gets executed with a ‘follow’ pattern, worker logs are also sent directly to stdout). In any case, I’m totally open to rename ‘displayLogs’ to ‘followLogs’ or ‘streamLogs’, thoughts? Maybe ‘grabLogs’ or ‘captureLogs’? Any particular choice?

Comment (mumoshu) [reply brigade-worker/src/job.ts]: Not a strong opinion, but `streamLogs` sounds good to me.

My mental model is that, `brig run` *follows* logs from the brigade worker. On the other hand the worker *streams* relevant logs to the worker. This change adds job pod logs to what's streamed by the worker. So, using `stream` works to me.

@dgkanatsios @radu-matei WDYT?

Btw, thanks a lot for your PR :) This had been a long-awaited feature to me!

Comment (dgkanatsios) [reply brigade-worker/src/job.ts]: agreed, `streamLogs` is a great choice, pushed a commit with this change. Thanks for the comment, hope this feature proves to be useful to you!

**Comment by dgkanatsios** [comment]
@vdice and @radu-matei issue with the linter is that it requires import alias for short or dash-containing package names like:
- `"k8s.io/api/core/v1"` should become `v1 "k8s.io/api/core/v1"`
- `"github.com/emicklei/go-restful"` should become `restful "github.com/emicklei/go-restful"`
- `"gopkg.in/yaml.v2"` should become `yaml "gopkg.in/yaml.v2"`
- `"gopkg.in/gin-gonic/gin.v1"` should become `gin "gopkg.in/gin-gonic/gin.v1"`

I can fix that and re-commit here (well, VS Code 'save-file' functionality fixes that), just wanna make sure this is the desired behaviour.

**Comment by vdice** [comment]
@dgkanatsios interesting; I wonder if this was a recent linting update/change?  SGTM.

**Comment by dgkanatsios** [comment]
thank you, just pushed, it seems OK now. As to how this was introduced, no clue, really. Maybe it comes with golangci? Have to admit that it 'feels' better, though

**Comment by vdice** [comment]
@radu-matei could you take a look when convenient?

**Comment by dgkanatsios** [comment]
thanks! about the commit history, I was scratching my head as well, not exactly sure why this happens. Branch on my fork ([link](https://github.com/dgkanatsios/brigade/tree/job-pod-logs)) mentions 'This branch is 6 commits ahead of Azure:master.' which is accurate...

File: brigade-worker/src/job.ts (hunk 1 above)
Comment (radu-matei) [reply brigade-worker/src/job.ts]: I know, too late, but `streamLogs` is a great name 😄

**Comment by radu-matei** [comment]
One question about the `phase Running/Succeeded` output - why can't we remove it when following the logs (or always?).

cc @mumoshu, who I think worked on the delegated runner.

File: brigade-worker/src/k8s.ts [hunk 2]
```diff
@@ -46,7 +46,7 @@ const wrapClient = fns => {
   // wrap client methods with retry logic
   for (let fn of fns) {
     let originalFn = defaultClient[fn.name];
-    defaultClient[fn.name] = function() {
+    defaultClient[fn.name] = function () {
```
Comment (radu-matei) [inline brigade-worker/src/k8s.ts]: I don't think there's an actual change here, right?
As it doesn't add anything, and we should probably add some linters for TypeScript and fix them everywhere (see #739), would it make sense to revert this type of changes in this file?

File: brigade-worker/src/k8s.ts [hunk 3]
```diff
@@ -439,7 +439,7 @@ export class JobRunner implements jobs.JobRunner {
     // appended to job name.
     return `${this.project.name.replace(/[.\/]/g, "-")}-${
       this.job.name
-    }`.toLowerCase();
+      }`.toLowerCase();
```
Comment (radu-matei) [inline brigade-worker/src/k8s.ts]: Here as well.

File: brigade-worker/src/k8s.ts [hunk 4]
```diff
@@ -461,7 +461,7 @@ export class JobRunner implements jobs.JobRunner {
    */
   public run(): Promise<jobs.Result> {
     return this.start()
-      .then(r => r.wait())
+      .then(r => this.wait())
```
Comment (radu-matei) [inline brigade-worker/src/k8s.ts]: Is this a linter recommendation?

```
Promise<JobRunner>.then<jobs.Result, never>(onfulfilled?: (value: jobs.JobRunner) => jobs.Result>
```

Technically, the `r` parameter passed to the promise is a `JobRunner`, so we can safely do `r.wait()` - I'm curious why the change, and if I'm missing something.

File: brigade-worker/src/k8s.ts [hunk 5]
```diff
@@ -626,11 +630,16 @@ export class JobRunner implements jobs.JobRunner {
           this.logger.log("Pod not yet scheduled");
           return;
         }
+
         let phase = this.pod.status.phase;
         if (phase == "Succeeded") {
           clearTimers();
           let result = new K8sResult(phase);
           resolve(result);
+        } else if (phase == "Running") { // make sure Pod is running before we start following its logs
```
Comment (radu-matei) [inline brigade-worker/src/k8s.ts]: nit: can the comment be on the previous line?
(This way, we're consistent with the single line comments - at least in this file).

File: brigade-worker/src/k8s.ts [hunk 6]
```diff
@@ -583,6 +583,8 @@ export class JobRunner implements jobs.JobRunner {
     return req;
   }
 
+
+
```
Comment (radu-matei) [inline brigade-worker/src/k8s.ts]: nit: One extra newline here.

File: brigade-worker/src/k8s.ts (hunk 5 above)
Comment (dgkanatsios) [reply brigade-worker/src/k8s.ts]: sure, fixing

File: brigade-worker/src/k8s.ts (hunk 6 above)
Comment (dgkanatsios) [reply brigade-worker/src/k8s.ts]: fixing

File: brigade-worker/src/k8s.ts (hunk 4 above)
Comment (dgkanatsios) [reply brigade-worker/src/k8s.ts]: nah, this was Dimitris-visual-linter :)
probably should have discussed this first, but it seemed to me that both Promises should have either _r.Function_ or _this.Function_. In any case, reverting this as well

File: brigade-worker/src/k8s.ts (hunk 2 above)
Comment (dgkanatsios) [reply brigade-worker/src/k8s.ts]: this was most certainly caused by VS Code auto-formatter on cmd-s while I was editing on Mac. I'm now editing on Windows and, weirdly, ctrl-s does not format it, so I'll revert it (and I'm pretty sure that I haven't done any explicit JS configuration on VS Code settings, so who knows...)

File: brigade-worker/src/k8s.ts (hunk 3 above)
Comment (dgkanatsios) [reply brigade-worker/src/k8s.ts]: same as above

**Comment by dgkanatsios** [comment]
thanks @radu-matei for the review, definitely +1 for the usage of TS linters
